Add RealtimeSanitizer to the CMake build system

sakertooth opened the pull request:
See title. More info on this code sanitizer and how to use it can be found [here](https://clang.llvm.org/docs/RealtimeSanitizer.html).

sakertooth committed:
Add rtsan

sakertooth committed:
Change order of creation

sakertooth committed:
Add LMMS_NONBLOCKING macro

sakertooth committed:
Apply LMMS_NONBLOCKING to audio device callbacks

sakertooth committed:
Apply LMMS_NONBLOCKING to render function

sakertooth committed:
Apply LMMS_NONBLOCKING macro to audio worker threads

sakertooth committed:
Use add_compile_definitions instead of setting LMMS_NONBLOCKING

sakertooth committed:
Allow rtsan to work in any build type

sakertooth committed:
Use empty attribute if WANT_RTSAN is not set

sakertooth committed:
Do some small refactoring

sakertooth committed:
Only apply LMMS_NONBLOCKING is more selective places

messmerd commented on CMakeLists.txt:
```diff
@@ -696,14 +697,23 @@ else()
 			# todo CMake 3.13: use add_compile_options/add_link_options instead
-			set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -fsanitize=${sanitizer} ${additional_flags}" PARENT_SCOPE)
-			set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -fsanitize=${sanitizer} ${additional_flags}" PARENT_SCOPE)
+			set(${C_FLAGS} "${${C_FLAGS}} -fsanitize=${sanitizer} ${additional_flags}" PARENT_SCOPE)
```
We require at least CMake 3.13 now, so maybe this would be a good time to fix the TODO comment on the line above?

messmerd commented on CMakeLists.txt:
```diff
@@ -712,12 +722,21 @@ function(add_sanitizer sanitizer supported_compilers want_flag status_flag)
+
+add_sanitizer(realtime "Clang" WANT_RTSAN STATUS_RTSAN any)
+
+if(WANT_RTSAN)
+	add_compile_definitions(LMMS_NONBLOCKING=[[clang::nonblocking]])
+else()
+	add_compile_definitions(LMMS_NONBLOCKING=[[]])
+endif()
```
I think this would be better in `lmmsconfig.h`. If you do it here, it's not clear in the code that `LMMS_NONBLOCKING` actually exists, so it may be undefined from the IDE's perspective and you may have to contend with red squiggly lines under functions that use it

sakertooth replied:
Fair concern, but I ran into issues doing it this way. My CMake knowledge is a bit janky so maybe I did something wrong. I'll try again.

JohannesLorenz replied:
What's the motivation to be able to turn it off? I would have thought that an unknown attribute is just getting ignored, so if the compile flags for the sanitizer are off, it should still compile fine with `[[clang::nonblocking]]`?

sakertooth replied:
Ah, I didn't know unknown attributes were ignored. I thought it would cause a compiler error. I'll simplify this down.

sakertooth replied:
Not sure if we should even use the `LMMS_NONBLOCKING` macro then, but I do like how it doesn't make compiler specific references to the main codebase.

sakertooth committed:
Apply suggestions from messmerd and some more

sakertooth committed:
Add noexcept along with LMMS_NONBLOCKING
Clang gives a warning (-Wperf-constraint-implies-noexcept) if noexcept is not added along with attributes like [[clang::nonblocking]]

JohannesLorenz reviewed (approved):
This is generally good.

Of course, this cannot fix the 1000s of expected findings and thus cannot add a CI.

Reminder that it might make sense to update the wiki (and then, maybe update submodule `doc/wiki` inside this PR?)

sakertooth commented on CMakeLists.txt:
```diff
@@ -701,9 +702,9 @@ function(add_sanitizer sanitizer supported_compilers want_flag status_flag)
-			set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -fsanitize=${sanitizer} ${additional_flags}" PARENT_SCOPE)
-			set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -fsanitize=${sanitizer} ${additional_flags}" PARENT_SCOPE)
+
+			add_compile_options(-fsanitize=${sanitizer} ${additional_flags})
```
Overkill I think since it depends on the compiler being used (e.g., Clang has AddressSanitizer, but MSVC does to, and both will most likely have introduced it to the compiler at different times, and the versioning simply doesn't translate to different compilers). However I do think I can refactor this function a bit still.

sakertooth committed:
Do mini refactoring of add_sanitizer function

sakertooth committed:
Update the wiki to explain RTsan

sakertooth committed:
Remove stoat instructions

sakertooth commented:
CI broke. Apparently you can't update the wiki directly within the PR, so I'll stash my recent changes and apply them to the wiki after merge.

sakertooth committed:
Revert "Remove stoat instructions"

This reverts commit 0482f8c7984553548397e551d07ead50ccbffb31.

sakertooth committed:
Revert "Update the wiki to explain RTsan"

This reverts commit ca0b64672fe67ce0c7409b5be1cf87c9b4a3658f.

sakertooth committed:
Merge remote-tracking branch 'upstream' into add-rtsan

messmerd commented:
Could you also put `LMMS_NONBLOCKING` on these methods?
- `Effect::processImpl`
- `Effect::processBypassedImpl`
- `Instrument::play`
- `Instrument::playNote`

These are audio processing methods that run on an audio thread and should ideally be real-time safe

sakertooth commented:
> Could you also put `LMMS_NONBLOCKING` on these methods?
> 
> - `Effect::processImpl`
> 
> - `Effect::processBypassedImpl`
> 
> - `Instrument::play`
> 
> - `Instrument::playNote`
> 
> 
> 
> These are audio processing methods that run on an audio thread and should ideally be real-time safe

The realtime attribute should be recursive. If it wasn't, there's no telling how many places we would need to put it, and as changes come along we would have to make sure that attribute is where it is all the time.

That's why I believe the Clang documentation states that you put the attribute at the entry points.

sakertooth commented:
I put the attribute on the audio device callbacks, the main render function, and the worker thread `run` function, so it should be applied where necessary, even on thread boundaries.

sakertooth commented:
I'll actually apply the macro to the FIFO writer thread instead of the main render function actually. The clang documentation expects it should be placed where threads with realtime execution start executing only.

messmerd commented:
@sakertooth 
> The realtime attribute should be recursive. If it wasn't, there's no telling how many places we would need to put it, and as changes come along we would have to make sure that attribute is where it is all the time.

The plugins are built separately though, so the compiler will never see the `[[clang::nonblocking]]` attribute when compiling them. And plugin developers should know that those methods they are expected to implement should be real-time safe.

sakertooth commented:
> @sakertooth 
> 
> > The realtime attribute should be recursive. If it wasn't, there's no telling how many places we would need to put it, and as changes come along we would have to make sure that attribute is where it is all the time.
> 
> 
> 
> The plugins are built separately though, so the compiler will never see the `[[clang::nonblocking]]` attribute when compiling them. And plugin developers should know that those methods they are expected to implement should be real-time safe.

Could you test this for me real quick? I don't have Clang 20 at the moment, but I'm pretty sure it applies to the plugins since the code will we are still executing is plugin code through the entry points.

And I think plugin implementers should know about real time safety from the get-go. If we put it there, then why not everywhere else? I understand the benefit of knowing what functions should be real time safe, but there is a huge chance we will miss functions that are supposed to be running in real time.

If we do this, then I find it's only right to apply it uniformly, which isn't really sustainable.

messmerd commented:
> Could you test this for me real quick? I don't have Clang 20 at the moment, but I'm pretty sure it applies to the plugins since the code will we are still executing is plugin code through the entry points.

I don't have Clang 20 either currently, but if the `[[clang::nonblocking]]` attribute is not visible to the plugin translation units, there is no way the compiler could know what is supposed to be real-time safe while building the plugin DLL/SO, unless the compiler stores that information in the object files or static library that the plugins are linked against or some other compiler magic, but I wouldn't count on it.

> And I think plugin implementers should know about real time safety from the get-go.

The entire point of using `[[clang::nonblocking]]` is so that the real-time requirements are explicitly stated and can be enforced. If we thought it was good enough that real-time safety was something people "should know about", this PR would not exist.

> If we put it there, then why not everywhere else?

The methods I mentioned are *the* main audio processing methods for all the plugins. There are few places better suited to add `LMMS_NONBLOCKING` than there.

And importantly, the `Effect` and `Instrument` classes are meant to form the basis of LMMS's plugin API. It is a *public* API used by plugin developers, not a hidden implementation detail. And in any good API, you must very clearly communicate how methods are expected to be used, and leverage the type system as much as you can to make it difficult to use incorrectly. The `[[clang::nonblocking]]` attribute gives us the opportunity for the first time to not just specify but also enforce that real-time requirement in our plugin API, so it would make no sense to not use it there.

The current plugin API is in very bad shape due to people treating plugins and the plugin API like they're part of LMMS core when they are supposed to be separate entities. This has made it difficult to tell where a plugin stops and the LMMS core begins which makes it hard to even call them "plugins". Ideally, a plugin developer should not need to read even a single line of code from LMMS core or know anything about it in order to write an LMMS plugin, because the plugin API should tell them everything they need to know. The fact that this is very much not the case is a massive failure IMO. Let's not make it worse.

sakertooth commented:
> I don't have Clang 20 either currently, but if the [[clang::nonblocking]] attribute is not visible to the plugin translation units, there is no way the compiler could know what is supposed to be real-time safe while building the plugin DLL/SO, unless the compiler stores that information in the object files or static library that the plugins are linked against or some other compiler magic, but I wouldn't count on it.

RTsan doesn't need to know AFAICT, it makes those decisions based on what functions are called in the call stack. We call the plugin functions on worker threads, so RTsan should definitely pick it up. I think for this situation, we need testing to make sure, otherwise it is a debate on theoreticals.

From Clang themselves:
> At run-time, if RTSan detects a call to malloc, free, pthread_mutex_lock, or anything else known to have a non-deterministic execution time in a function marked [[clang::nonblocking]] it raises an error.
---

> The entire point of using [[clang::nonblocking]] is so that the real-time requirements are explicitly stated and can be enforced. If we thought it was good enough that real-time safety was something people "should know about", this PR would not exist.

It serves two purposes: one is to raise an error on a violation of real-time safety when executing a function, and yes it also effectively serves as an indicator that this is a real-time function, but the main point is to catch real-time violations more easily. That's always been the point. The PR would still exist to serve that main purpose.

However, I do not agree with putting `[[clang::nonblocking]]` in only those specific spots you listed. That decision does not seem like a cohesive one and completely ignores everything else that *should* be marked for real-time safety. For the same reasons you want `LMMS_NONBLOCKING` to be applied for those functions, the other ones as well (like `Song::processNextBuffer`, `Sample::play`, tons of them..) are perfect candidates given that reasoning.

> And importantly, the Effect and Instrument classes are meant to form the basis of LMMS's plugin API. It is a public API used by plugin developers, not a hidden implementation detail. And in any good API, you must very clearly communicate how methods are expected to be used, and leverage the type system as much as you can to make it difficult to use incorrectly. The [[clang::nonblocking]] attribute gives us the opportunity for the first time to not just specify but also enforce that real-time requirement in our plugin API, so it would make no sense to not use it there.

But you have not given a good reason why only those functions. As I've mentioned, it is only fair to respect the public API everywhere else in the code, but this isn't sustainable long term unless we support a more general interface (i.e., specifying it in an abstract class that everyone else inherits such details from). What I see happening is a mixture of the use of `LMMS_NONBLOCKING` in some places, and other places, not so much. I don't like the idea of that inconsistency.

> The current plugin API is in very bad shape due to people treating plugins and the plugin API like they're part of LMMS core when they are supposed to be separate entities. This has made it difficult to tell where a plugin stops and the LMMS core begins which makes it hard to even call them "plugins". Ideally, a plugin developer should not need to read even a single line of code from LMMS core or know anything about it in order to write an LMMS plugin, because the plugin API should tell them everything they need to know. The fact that this is very much not the case is a massive failure IMO. Let's not make it worse.

I agree. I just don't think I agree with the over-emphasis on applying it in one specific spot in the code where it has no functional effect but to signal to developers that the function is real-time safe. The space that involves applying it to only the specific entry-points of real-time execution (where this PR puts them) is limited and manageable. However, for places that we want to mark as real-time safe? The surface area for where it should be applied increases to a point where I don't even see it being sustainable long term as the APIs change.

---

I wont have time to work on LMMS anymore, so if you want the changes you can add them to the branch. There's no point in debating over this. Note that I do think it is important and useful to signal that the function is real-time safe, but having it *just for the plugin API* for that reason in particular isn't "complete" as an idea. It will most likely lead to inconsistent habits and result in that macro being in places where it shouldn't be out of feeling (e.g. functions that are called by other real-time functions that aren't even in the public API for a certain class).

messmerd commented:
> RTsan doesn't need to know AFAICT, it makes those decisions based on what functions are called in the call stack. We call the plugin functions on worker threads, so RTsan should definitely pick it up.

That might be true when compiling the LMMS core, but when compiling a plugin, the compiler has no way of knowing that somewhere up the call stack the plugin's process method will be called by a function marked `[[clang::nonblocking]]` because that information does not exist anywhere in the plugin's source files or included headers. It probably doesn't even know that the process method will be called at all. The only way for the compiler to learn that information is if it stored it somehow when compiling the LMMS core and can check it at link time, but I would not count on this.

> However, I do not agree with putting [[clang::nonblocking]] in only those specific spots you listed. That decision does not seem like a cohesive one and completely ignores everything else that should be marked for real-time safety.

I'm not advocating putting it everywhere haphazardly. Only places where it is needed. The process methods are by far the most important methods of a plugin and (to my knowledge) the only real-time methods that plugins implement, which makes them an obvious choice to apply the attribute.

> For the same reasons you want LMMS_NONBLOCKING to be applied for those functions, the other ones as well (like Song::processNextBuffer, Sample::play, tons of them..) are perfect candidates given that reasoning.

None of those examples are part of LMMS's public API though. A plugin developer does not need to know they exist.

> But you have not given a good reason why only those functions. As I've mentioned, it is only fair to respect the public API everywhere else in the code, but this isn't sustainable long term unless we support a more general interface (i.e., specifying it in an abstract class that everyone else inherits such details from). 

- It's a public API, and in any good public API, all requirements need to be clearly communicated to the people using it (plugin developers).
- There is no "public API everywhere else in the code". The plugin API is the *only* public API that LMMS has (except maybe the `src/common/` stuff), so applying the nonblocking attributes there is not incohesive or random. All other APIs are internal to the LMMS application.
- RtSan static analysis should be enabled for plugins. Without marking the real-time methods with `[[clang::nonblocking]]`, the compiler will likely be unable to infer that the process methods are supposed to be real-time safe when trying to compile a plugin.

> I just don't think I agree with the over-emphasis on applying it in one specific spot in the code where it has no functional effect but to signal to developers that the function is real-time safe. The space that involves applying it to only the specific entry-points of real-time execution (where this PR puts them) is limited and manageable.

Like I said, the plugins and LMMS are two separate entities. The plugin process methods are the entry-points of real-time execution for plugins.

> I wont have time to work on LMMS anymore, so if you want the changes you can add them to the branch.

I will be gone for a while too (going on vacation for a couple weeks starting next week). You've done a good job on this PR so far and I'd like to test it out once I have the time, though it may be a bit.

sakertooth commented:
> RtSan static analysis should be enabled for plugins. Without marking the real-time methods with [[clang::nonblocking]], the compiler will likely be unable to infer that the process methods are supposed to be real-time safe when trying to compile a plugin.

Apologies for the misunderstanding, I was under the assumption that RTSan only had use at runtime, not compile-time. I think I am now in agreement with putting it there.

From Clang:

> RTSan performs its analysis at run-time but shares the [[clang::nonblocking]] attribute with the [Function Effect Analysis](https://clang.llvm.org/docs/FunctionEffectAnalysis.html) system, which operates at compile-time to detect potential real-time safety violations. For comprehensive detection of real-time safety issues, it is recommended to use both systems together.

It's very evident that I did not read this full paragraph, or skimmed over it.

> None of those examples are part of LMMS's public API though. A plugin developer does not need to know they exist.

I was thinking of the other `public:`, apologies for the misunderstanding.

> Like I said, the plugins and LMMS are two separate entities. The plugin process methods are the entry-points of real-time execution for plugins.

I incorrectly assumed RTSan was only useful at run-time, but I still think if LMMS were to run normally, RTSan would catch violations from within plugins at run-time. I don't necessarily see how they RTsan wouldn't catch the violations at run-time any other way unless the process methods were called elsewhere, maybe in a distinct program, but I could be wrong/don't see this happening anytime soon.

It also does not seem like the static analysis goes through all real-time function paths for any given function marked `nonblocking`. Instead, it will just flag the ones called in the function not marked `nonblocking` as a problem. So maybe the goal *is* to eventually apply it to all real-time function paths after we make them real-time, but it would still stand that runtime analysis is better at catching violations than static analysis.

So, if I understand this correctly, the main problem mentioned is that RTSan's static analysis won't have any impact regarding plugins since they are compiled separately. The confusing part there is that I can say that for many real-time functions that weren't marked real-time that are compiled and linked into `lmmsobjs`. So the only real benefit I see is that since the plugin API is meant to be a bit more separate from all of LMMS, the macro mainly serves to communicate to developers that the function is supposed to run in real-time.

Overall though, I am a bit confused about the whole thing, so I'll have to postpone this a couple months late if anything.

> You've done a good job on this PR

Thanks.

sakertooth commented:
Turns out I am inevitably going to have to apply `LMMS_NONBLOCKING` to all real-time functions. Attributes aren't inherited from the base class from what I've researched, so its easier to ensure it is applied everywhere where necessary.

sakertooth commented:
I'll be picking this back up (given I have free time).

sakertooth commented:
> RTsan doesn't need to know AFAICT, it makes those decisions based on what functions are called in the call stack. We call the plugin functions on worker threads, so RTsan should definitely pick it up.

> That might be true when compiling the LMMS core, but when compiling a plugin, the compiler has no way of knowing that somewhere up the call stack the plugin's process method will be called by a function marked [[clang::nonblocking]] because that information does not exist anywhere in the plugin's source files or included headers. It probably doesn't even know that the process method will be called at all. The only way for the compiler to learn that information is if it stored it somehow when compiling the LMMS core and can check it at link time, but I would not count on this.

@messmerd, do we at least agree that RTsan *can* detect non RT safe code from within plugins at runtime? If not, I did a test that injected some memory allocation code into `Amplifier::processImpl`. RTsan can detect these just fine, so the only worry should be if it can do at compile time.

```bash
SUMMARY: RealtimeSanitizer: unsafe-library-call (/home/saker/projects/lmms/build/lmms+0x2612a7) (BuildId: cac588ea9d96ca134f551712dc3eb71e43f921f1) in free
==13968==ERROR: RealtimeSanitizer: unsafe-library-call
Intercepted call to real-time unsafe function `malloc` in real-time context!
    #0 0x5654d1676984 in malloc (/home/saker/projects/lmms/build/lmms+0x261984) (BuildId: cac588ea9d96ca134f551712dc3eb71e43f921f1)
    #1 0x7f45e50b6f44 in operator new(unsigned long) /usr/src/debug/gcc/gcc/libstdc++-v3/libsupc++/new_op.cc:50:22
    #2 0x7f45bc3bf224 in lmms::AmplifierEffect::processImpl(lmms::SampleFrame*, unsigned long) /home/saker/projects/lmms/plugins/Amplifier/Amplifier.cpp:63:11
    #3 0x5654d17422c3 in lmms::Effect::processAudioBuffer(lmms::AudioBuffer&) /home/saker/projects/lmms/src/core/Effect.cpp:120:22
    #4 0x5654d17445c4 in lmms::EffectChain::processAudioBuffer(lmms::AudioBuffer&) /home/saker/projects/lmms/src/core/EffectChain.cpp:200:26
    #5 0x5654d174ea65 in lmms::MixerChannel::doProcessing() /home/saker/projects/lmms/src/core/Mixer.cpp:213:30
    #6 0x5654d16e93de in lmms::ThreadableJob::process() /home/saker/projects/lmms/include/ThreadableJob.h:77:4
    #7 0x5654d16e893f in lmms::AudioEngineWorkerThread::JobQueue::run() /home/saker/projects/lmms/src/core/AudioEngineWorkerThread.cpp:87:10
    #8 0x5654d16e8c0e in lmms::AudioEngineWorkerThread::startAndWaitForJobs() /home/saker/projects/lmms/src/core/AudioEngineWorkerThread.cpp:160:17
    #9 0x5654d1751864 in lmms::Mixer::masterMix(lmms::SampleFrame*) /home/saker/projects/lmms/src/core/Mixer.cpp:709:3
    #10 0x5654d16e251c in lmms::AudioEngine::renderStageMix() /home/saker/projects/lmms/src/core/AudioEngine.cpp:304:9
    #11 0x5654d16e266f in lmms::AudioEngine::renderNextPeriod() /home/saker/projects/lmms/src/core/AudioEngine.cpp:328:2
    #12 0x5654d17edfdb in void lmms::AudioEngine::renderNextBuffer<lmms::InterleavedBufferView<float, (unsigned short)65535>>(lmms::InterleavedBufferView<float, (unsigned short)65535>) /home/saker/projects/lmms/include/AudioEngine.h:317:40
    #13 0x5654d17ed256 in lmms::AudioEngine::renderNextBuffer(lmms::InterleavedBufferView<float, (unsigned short)65535>) /home/saker/projects/lmms/include/AudioEngine.h:287:60
    #14 0x5654d1807787 in lmms::AudioSdl::sdlAudioCallback(unsigned char*, int) /home/saker/projects/lmms/src/core/audio/AudioSdl.cpp:174:17
    #15 0x5654d18074e5 in lmms::AudioSdl::sdlAudioCallback(void*, unsigned char*, int) /home/saker/projects/lmms/src/core/audio/AudioSdl.cpp:159:9
```

However, I dont find the compile time variant reasonable to be honest. The problem is that with it enabled (with `-Wfunction-effects` I believe), it gives warnings for functions marked `[[clang::nonblocking]]` that call functions not marked `[[clang::nonblocking]]`, and that's it.
```cpp
/home/saker/projects/lmms/include/AudioSoundIo.h:125:10: warning: function with 'nonblocking' attribute must not call non-'nonblocking' function 'lmms::AudioSoundIo::writeCallback' [-Wfunction-effects]
  125 |                 return ((AudioSoundIo *)outstream->userdata)->writeCallback(frame_count_min, frame_count_max);
      |                        ^
/home/saker/projects/lmms/include/AudioSoundIo.h:118:7: note: declaration cannot be inferred 'nonblocking' because it has no definition in this translation unit
  118 |         void writeCallback(int frame_count_min, int frame_count_max);
      |              ^
/home/saker/projects/lmms/include/AudioSoundIo.h:130:10: warning: function with 'nonblocking' attribute must not call non-'nonblocking' function 'lmms::AudioSoundIo::errorCallback' [-Wfunction-effects]
  130 |                 return ((AudioSoundIo *)outstream->userdata)->errorCallback(err);
      |                        ^
/home/saker/projects/lmms/include/AudioSoundIo.h:119:7: note: declaration cannot be inferred 'nonblocking' because it has no definition in this translation unit
  119 |         void errorCallback(int err);
      |              ^
/home/saker/projects/lmms/include/AudioSoundIo.h:135:10: warning: function with 'nonblocking' attribute must not call non-'nonblocking' function 'lmms::AudioSoundIo::underflowCallback' [-Wfunction-effects]
  135 |                 return ((AudioSoundIo *)outstream->userdata)->underflowCallback();
      |                        ^
/home/saker/projects/lmms/include/AudioSoundIo.h:120:7: note: declaration cannot be inferred 'nonblocking' because it has no definition in this translation unit
  120 |         void underflowCallback();
      |              ^
```

It doesn't go through the entire call stack (which makes sense), so to make good use of it, *all RT functions* have to be marked. And then for the functions that are called by them, if they are meant to run in real-time, they too have to be marked. I don't understand how this is not infeasible. Especially with new features on the way like audio recording, focusing on just marking the actual RT entry points (i.e., the `QThread::run` or audio callback for RT audio threads) is much more doable.

This isn't even mentioning the third party libraries that the audio threads call (e.g. `libsamplerate`). If we really wanted to be thorough, we would also have to find a way to mark those functions with `[[clang::nonblocking]]` so it can be picked up at compile time too.

Yes it can make sense for the public API to explicitly mention real-time constraints, but we would have to cover a lot more functions. This is error prone and there are no guarantees we will cover all the ground we need to cover. And given that developers are generally familiar with such real-time constraints, I rather not.

---

It can detect non RT-safe code from our plugins at run-time, so I am not worried anymore. I think that's enough, and I'm also not convinced we can make good use of the compile-time variant.

messmerd commented:
> It doesn't go through the entire call stack (which makes sense), so to make good use of it, _all RT functions_ have to be marked. And then for the functions that are called by them, if they are meant to run in real-time, they too have to be marked. I don't understand how this is not infeasible.

Ok, we could worry about getting compile-time analysis with `-Wfunction-effects` working sometime in the future, and just focus on runtime now.

Getting `-Werror -Wfunction-effects` to compile would require a lot of work marking functions, but more significantly, we'd actually have to fix all of the many real-time safety violations that currently exist.

> do we at least agree that RTsan _can_ detect non RT safe code from within plugins at runtime?

Yes, it should detect real-time violations in plugins at runtime, even if `processImpl` is not marked with `LMMS_NONBLOCKING`.

But even if we're just focusing on runtime sanitization and leaving compile-time analysis for the future, I think it's a good idea to mark plugin process methods as `LMMS_NONBLOCKING` too.

> Yes it can make sense for the public API to explicitly mention real-time constraints, but we would have to cover a lot more functions.

It's just 4 more lines of code to change (the `Effect::processImpl`, `Effect::processBypassedImpl`, `Instrument::play`, and `Instrument::playNote` declarations), and then our plugin API documents the real-time safety requirements. The `[[clang::nonblocking]]` attribute when applied to a virtual method declaration is inherited by derived class methods that override it, so you wouldn't need to apply it to every plugin implementation - only the base class.

sakertooth commented:
> Getting `-Werror -Wfunction-effects` to compile would require a lot of work marking functions, but more significantly, we'd actually have to fix all of the many real-time safety violations that currently exist.

This is made significantly more difficult with 3rd party libraries the audio threads call into (e.g. Qt). We would have to create thin annotated wrappers for these libraries, which is a time sink. There's not much to gain from doing this when runtime detection is available to us. I don't even think the standard library has these functions annotated.

The compile-time RT evaluation component to `[[clang::nonblocking]]` is hard to take full advantage of, from what I've seen.

> It's just 4 more lines of code to change (the `Effect::processImpl`, `Effect::processBypassedImpl`, `Instrument::play`, and `Instrument::playNote` declarations), and then our plugin API documents the real-time safety requirements.

Functionally, I dont really see that much of a gain by doing this. It might help? But I don't really see it. Basic documentation also has the same effect barring compile-time evaluation of `[[clang::nonblocking]]`.

I suppose you would have a preference for these classes in particular as you've been doing work to improve the plugin API, which is great. But for the same reasons we apply it here, it ideally should be applied to all of the functions that are at the real-time boundary, not just the ones listed. Its just that the boundary is poorly defined right now, (with business objects interacting with the audio threads, sometimes even with the GUI) so the scope of which the attribute applies is needlessly large.

For any developers working on the audio engine or near the RT boundary, having `[[clang::nonblocking]]` on whatever function the audio threads execute would be really helpful, but how are we supposed to cover all of that ground and make sure everything is annotated properly? I don't have the time for that.

I can annotate the `Instrument` and `Effect` functions you gave, but I dont think I will be okay with it deep down, at least not right now. I'm not convinced of an actual benefit when compared to regular documentation. It seems a bit redundant. 
I guess I am looking for a hard real-time boundary in the code to go "this is where we should apply this attribute". A better boundary would imply a smaller, more controlled one, so annotating these functions right now somewhat feels like we didn't do the necessary step of creating that boundary first.

sakertooth commented:
> The [[clang::nonblocking]] attribute when applied to a virtual method declaration is inherited by derived class methods that override it, so you wouldn't need to apply it to every plugin implementation - only the base class.

Ideally `noexcept` would be applied to, but that doesn't automatically propagate the same way as `[[clang::nonblocking]]`, so I would need to update all of them.

messmerd commented:
> > The [[clang::nonblocking]] attribute when applied to a virtual method declaration is inherited by derived class methods that override it, so you wouldn't need to apply it to every plugin implementation - only the base class.
> 
> Ideally `noexcept` would be applied to, but that doesn't automatically propagate the same way as `[[clang::nonblocking]]`, so I would need to update all of them.

Oh true, that would be a pain. If you don't want to add `LMMS_NONBLOCKING` to the plugin API yet, I'm fine with doing it some other time. I plan to modify those methods anyway for the new plugin API.

sakertooth commented:
Slightly annoying, but it's done. Reviewing this is will probably be a lot harder now, but hopefully it isn't that big of an issue. CI might also scream at me but we'll see.

sakertooth commented:
From [Clang](https://clang.llvm.org/docs/RealtimeSanitizer.html) themselves:
> Typically, these attributes should be added onto the functions that are entry points for threads with real-time priority. These threads are subject to a fixed callback time, such as audio callback threads or rendering loops in video game code.

This is what I originally did.

I will retract adding it to plugins based on what Clang mentioned and also the simplicity of the PR. If you want to do it later, I won't stop you, but it is an oddly selective set of functions you want to add it to.

sakertooth committed:
Merge remote-tracking branch 'upstream' into add-rtsan
